File: Sandbox/MaiAT3PlusDecoder/src/base/MaiBitReader.cc
#include "SUB/MaiBitReader.h"
#include <stdio.h>

MaiBitReader::MaiBitReader(Mai_I32 byte_buf_size, Mai_I32 type)
{
	quene_in = new MaiQueue0(byte_buf_size);
	buffer = 0;
	bits_num = 0;
	this->type = type;
}

MaiBitReader::~MaiBitReader()
{
	mcs0.enter();

	delete quene_in;

	mcs0.leave();
}

Mai_Status MaiBitReader::moreByte()
{
	if (quene_in->GetLength())
	{
		Mai_I8 temp;
		quene_in->Out(&temp, 1);
		if (type == MaiBitReaderTypeHigh)
			buffer = (buffer << 8) | (temp & 0xFF);
		else if (type == MaiBitReaderTypeLow)
			buffer |= (temp & 0xFF) << bits_num;
		bits_num += 8;
		return 0;
	}
	else return -1;
}

Mai_Status MaiBitReader::addData(Mai_I8* src, Mai_I32 len_s)
{
	mcs0.enter();
	
	printf("addData(%d) : ", len_s);
	for (int n = 0; n < len_s; n++) printf("%02X-", (unsigned char)src[n]);
	printf("\n");

	quene_in->In(src, len_s);

	mcs0.leave();

	return 0;
}

Mai_I32 MaiBitReader::getRemainingBitsNum()
{
	mcs0.enter();

	Mai_I32 bits_remain = (quene_in->GetLength() << 3) + bits_num;

	mcs0.leave();

	return bits_remain;
}

Mai_I32 MaiBitReader::getWithI32Buffer(Mai_I32 bnum, Mai_Bool get_then_del_in_buf)
{
	mcs0.enter();

	while (bnum > bits_num) if (moreByte()) break;
	if (bnum <= bits_num)
	{
		Mai_I32 to_out = 0;

		if (type == MaiBitReaderTypeHigh)
			to_out = (buffer >> (bits_num - bnum)) & ((1 << bnum) - 1);
		else if (type == MaiBitReaderTypeLow)
			to_out = buffer & ((1 << bnum) - 1);

		if (get_then_del_in_buf)
		{
			bits_num -= bnum;

			if (type == MaiBitReaderTypeHigh)
				buffer = buffer & ((1 << bits_num) - 1);
			else if (type == MaiBitReaderTypeLow)
				buffer = (buffer >> bnum) & ((1 << bits_num) - 1);
		}

		mcs0.leave();
		printf("getWithI32Buffer(%d) : %d\n", bnum, to_out);
		return to_out;
	}
	else
	{
		mcs0.leave();
		return 0;
	}
}

File: Hle/CSPspEmu.Hle.Media/audio/At3/SUB/MaiBitReader.cs
//#define BIT_READER_THREAD_SAFE
//#define DEBUG_BIT_READER

using System;

namespace CSPspEmu.Hle.Media.audio.At3.SUB
{
    public sealed unsafe class MaiBitReader : IDisposable
    {
        public const int MaiBitReaderTypeHigh = 0;
        public const int MaiBitReaderTypeLow = 1;

        MaiQueue0 quene_in;
        int _buffer;
        int _bitsNum;
        int type;

        public MaiBitReader(int byteBufSize, int type = MaiBitReaderTypeHigh)
        {
            quene_in = new MaiQueue0(byteBufSize);
            _buffer = 0;
            _bitsNum = 0;
            this.type = type;
        }

        public void Dispose()
        {
#if BIT_READER_THREAD_SAFE
			lock (this)
#endif
            {
                quene_in.Dis();
            }
        }

        public bool MoreByte()
        {
            if (quene_in.GetLength() != 0)
            {
                var temp = new byte[1];
                quene_in.Out(temp, 1);
                if (type == MaiBitReaderTypeHigh)
                {
                    _buffer = (_buffer << 8) | (temp[0] & 0xFF);
                }
                else if (type == MaiBitReaderTypeLow)
                {
                    _buffer |= (temp[0] & 0xFF) << _bitsNum;
                }
                _bitsNum += 8;
                return false;
            }
            else
            {
                return true;
            }
        }

        public int AddData(byte* src, int lenS)
        {
            int ret = _addData(src, lenS);
#if DEBUG_BIT_READER
			Console.WriteLine("addData({0}) : {1}", len_s, BitConverter.ToString(src.GetArray(len_s)));
#endif
            return ret;
        }

        private int _addData(byte* src, int lenS)
        {
#if BIT_READER_THREAD_SAFE
			lock (this)
#endif
            {
                quene_in.In(src, lenS);
            }

            return 0;
        }

        public int GetRemainingBitsNum()
        {
#if BIT_READER_THREAD_SAFE
			lock (this)
#endif
            {
                int bitsRemain = (quene_in.GetLength() << 3) + _bitsNum;

                return bitsRemain;
            }
        }

        ////[MethodImpl(MethodImplOptions.AggressiveInlining)]
        public int GetWithI32Buffer(int bnum, bool getThenDelInBuf = true)
        {
            var value = _getWithI32Buffer(bnum, getThenDelInBuf);
#if DEBUG_BIT_READER
			Console.WriteLine("MaiBitReader.getWithI32Buffer({0}) : {1}", bnum, value);
#endif
            return value;
        }

        ////[MethodImpl(MethodImplOptions.AggressiveInlining)]
        private int _getWithI32Buffer(int bnum, bool getThenDelInBuf = true)
        {
#if BIT_READER_THREAD_SAFE
			lock (this)
#endif
            {
                while (bnum > _bitsNum)
                {
                    if (MoreByte()) break;
                }

                if (bnum <= _bitsNum)
                {
                    int toOut = 0;

                    if (type == MaiBitReaderTypeHigh)
                    {
                        toOut = (_buffer >> (_bitsNum - bnum)) & ((1 << bnum) - 1);
                    }
                    else if (type == MaiBitReaderTypeLow)
                    {
                        toOut = _buffer & ((1 << bnum) - 1);
                    }

                    if (getThenDelInBuf)
                    {
                        _bitsNum -= bnum;

                        if (type == MaiBitReaderTypeHigh)
                        {
                            _buffer = _buffer & ((1 << _bitsNum) - 1);
                        }
                        else if (type == MaiBitReaderTypeLow)
                        {
                            _buffer = (_buffer >> bnum) & ((1 << _bitsNum) - 1);
                        }
                    }

                    return toOut;
                }
                else
                {
                    return 0;
                }
            }
        }
    }
}